src/Dependencies/Collections/SegmentedArray.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Collections.Internal;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal static class SegmentedArray
+    {
+        /// <seealso cref="Array.Clear(Array, int, int)"/>
+        internal static void Clear<T>(SegmentedArray<T> array, int index, int length)
+        {
+            foreach (var memory in array.GetSegments(index, length))
+            {
+                memory.Span.Clear();
+            }
+        }
+
+        /// <seealso cref="Array.Copy(Array, Array, int)"/>
+        internal static void Copy<T>(SegmentedArray<T> sourceArray, SegmentedArray<T> destinationArray, int length)
+        {
+            if (length == 0)
+                return;
+
+            if ((uint)length <= (uint)sourceArray.Length
+                && (uint)length <= (uint)destinationArray.Length)
+            {
+                var sourceSegmentEnumerable = sourceArray.GetSegments(0, length);
+                var destinationSegmentEnumerable = destinationArray.GetSegments(0, length);
+
+                using var sourceSegmentEnumerator = sourceSegmentEnumerable.GetEnumerator();
+                using var destinationSegmentEnumerator = destinationSegmentEnumerable.GetEnumerator();
+                while (sourceSegmentEnumerator.MoveNext())
+                {
+                    destinationSegmentEnumerator.MoveNext();
+                    sourceSegmentEnumerator.Current.CopyTo(destinationSegmentEnumerator.Current);
+                }
+
+                return;
+            }
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+            if (length > sourceArray.Length)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            if (length > destinationArray.Length)
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+
+            throw new InvalidOperationException("This program location is thought to be unreachable.");
+        }
+
+        private static SegmentEnumerable<T> GetSegments<T>(this SegmentedArray<T> array, int offset, int length)
+            => new(array, offset, length);
+
+        private readonly struct SegmentEnumerable<T> : IEnumerable<Memory<T>>
```

Comment:
>IEnumerable<Memory<T>> [](start = 55, length = 22)

Why bother actually implementing `IEnumerable`/`IEnumerator` on these structs? Unless it's coming in a commit I haven't gotten to yet, it doesn't appear that you're actually exposing them via these interfaces anywhere?

Reply:
I wish we had shapes. For now I can remove the interface implementations.